File: android/app/src/main/java/com/gooseco/myliftsquad/data/api/ShareApiService.kt
package com.gooseco.myliftsquad.data.api

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject
import java.util.concurrent.TimeUnit

data class AthleteRef(val name: String, val slug: String)
data class SharedSquad(val name: String, val athletes: List<AthleteRef>)
data class SharedBundle(val squads: List<SharedSquad>)

class ShareApiService {

    private val client = OkHttpClient.Builder()
        .connectTimeout(15, TimeUnit.SECONDS)
        .readTimeout(15, TimeUnit.SECONDS)
        .build()

    private val baseUrl = "https://myliftsquad-api.gooseyirl.workers.dev"
    private val json = "application/json; charset=utf-8".toMediaType()

    // ── Single squad ─────────────────────────────────────────────────

    suspend fun shareSquad(name: String, athletes: List<AthleteRef>): String =
        withContext(Dispatchers.IO) {
            val body = buildSquadJson(name, athletes).toString().toRequestBody(json)
            val request = Request.Builder().url("$baseUrl/squads").post(body).build()
            val response = client.newCall(request).execute()
            val responseBody = response.body?.string() ?: throw Exception("Empty response")
            if (!response.isSuccessful) throw Exception("Server error: ${response.code}")
            JSONObject(responseBody).getString("code")
        }

    /** Returns null if the code was not found (404), throws on other errors. */
    suspend fun tryImportSquad(code: String): SharedSquad? =
        withContext(Dispatchers.IO) {
            val request = Request.Builder()
                .url("$baseUrl/squads/${code.uppercase()}")
                .get().build()
            val response = client.newCall(request).execute()
            if (response.code == 404) return@withContext null
            val responseBody = response.body?.string() ?: throw Exception("Empty response")
            if (!response.isSuccessful) throw Exception("Server error: ${response.code}")
            parseSharedSquad(JSONObject(responseBody))
        }

    // ── Bundle (multiple squads) ──────────────────────────────────────

    suspend fun shareBundle(squads: List<SharedSquad>): String =
        withContext(Dispatchers.IO) {
            val body = JSONObject().apply {
                put("squads", JSONArray().apply {
                    squads.forEach { squad ->
                        put(buildSquadJson(squad.name, squad.athletes))
                    }
                })
            }.toString().toRequestBody(json)

            val request = Request.Builder().url("$baseUrl/bundles").post(body).build()
            val response = client.newCall(request).execute()
            val responseBody = response.body?.string() ?: throw Exception("Empty response")
            if (!response.isSuccessful) throw Exception("Server error: ${response.code}")
            JSONObject(responseBody).getString("code")
        }

    /** Returns null if the code was not found (404), throws on other errors. */
    suspend fun tryImportBundle(code: String): SharedBundle? =
        withContext(Dispatchers.IO) {
            val request = Request.Builder()
                .url("$baseUrl/bundles/${code.uppercase()}")
                .get().build()
            val response = client.newCall(request).execute()
            if (response.code == 404) return@withContext null
            val responseBody = response.body?.string() ?: throw Exception("Empty response")
            if (!response.isSuccessful) throw Exception("Server error: ${response.code}")

            val obj = JSONObject(responseBody)
            val squadsArr = obj.getJSONArray("squads")
            val squads = (0 until squadsArr.length()).map { i ->
                parseSharedSquad(squadsArr.getJSONObject(i))
            }
            SharedBundle(squads = squads)
        }

    // ── Helpers ───────────────────────────────────────────────────────

    private fun buildSquadJson(name: String, athletes: List<AthleteRef>): JSONObject =
        JSONObject().apply {
            put("name", name)
            put("athletes", JSONArray().apply {
                athletes.forEach { a ->
                    put(JSONObject().apply {
                        put("name", a.name)
                        put("slug", a.slug)
                    })
                }
            })
        }

    private fun parseSharedSquad(obj: JSONObject): SharedSquad {
        val name = obj.getString("name")
        val athletesArr = obj.getJSONArray("athletes")
        val athletes = (0 until athletesArr.length()).map { i ->
            val a = athletesArr.getJSONObject(i)
            AthleteRef(name = a.getString("name"), slug = a.getString("slug"))
        }
        return SharedSquad(name = name, athletes = athletes)
    }
}

File: ios/MyLiftSquad/Sources/Data/API/ShareApiService.swift
import Foundation

struct AthleteRef: Codable {
    let name: String
    let slug: String
}

struct SharedSquad: Codable {
    let name: String
    let athletes: [AthleteRef]
}

struct SharedBundle: Decodable {
    let squads: [SharedSquad]
}

actor ShareApiService {
    static let shared = ShareApiService()
    private let baseURL: String
    private let session: URLSession

    private init() {
        self.baseURL = "https://myliftsquad-api.gooseyirl.workers.dev"
        self.session = .shared
    }

    init(baseURL: String, session: URLSession) {
        self.baseURL = baseURL
        self.session = session
    }

    // ── Single squad ──────────────────────────────────────────────────

    func shareSquad(name: String, athletes: [AthleteRef]) async throws -> String {
        guard let url = URL(string: "\(baseURL)/squads") else { throw URLError(.badURL) }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONEncoder().encode(SharedSquad(name: name, athletes: athletes))

        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse, http.statusCode == 201 else {
            throw URLError(.badServerResponse)
        }
        struct CodeResponse: Decodable { let code: String }
        return try JSONDecoder().decode(CodeResponse.self, from: data).code
    }

    /// Returns nil if the code was not found (404), throws on other errors.
    func tryImportSquad(code: String) async throws -> SharedSquad? {
        guard let url = URL(string: "\(baseURL)/squads/\(code.uppercased())") else { throw URLError(.badURL) }
        let (data, response) = try await session.data(from: url)
        guard let http = response as? HTTPURLResponse else { throw URLError(.badServerResponse) }
        if http.statusCode == 404 { return nil }
        guard http.statusCode == 200 else { throw URLError(.badServerResponse) }
        return try JSONDecoder().decode(SharedSquad.self, from: data)
    }

    // ── Bundle (multiple squads) ──────────────────────────────────────

    func shareBundle(squads: [SharedSquad]) async throws -> String {
        guard let url = URL(string: "\(baseURL)/bundles") else { throw URLError(.badURL) }
        struct Payload: Encodable { let squads: [SharedSquad] }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONEncoder().encode(Payload(squads: squads))

        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse, http.statusCode == 201 else {
            throw URLError(.badServerResponse)
        }
        struct CodeResponse: Decodable { let code: String }
        return try JSONDecoder().decode(CodeResponse.self, from: data).code
    }

    /// Returns nil if the code was not found (404), throws on other errors.
    func tryImportBundle(code: String) async throws -> SharedBundle? {
        guard let url = URL(string: "\(baseURL)/bundles/\(code.uppercased())") else { throw URLError(.badURL) }
        let (data, response) = try await session.data(from: url)
        guard let http = response as? HTTPURLResponse else { throw URLError(.badServerResponse) }
        if http.statusCode == 404 { return nil }
        guard http.statusCode == 200 else { throw URLError(.badServerResponse) }
        return try JSONDecoder().decode(SharedBundle.self, from: data)
    }
}
